src/main/java/com/ryanharter/auto/value/moshi/AutoValueMoshiExtension.java

```diff
@@ -217,28 +246,42 @@ MethodSpec generateConstructor(Map<String, TypeName> properties) {
     return types;
   }
 
-  public MethodSpec createJsonAdapterMethod(TypeName autoValueClass, TypeSpec jsonAdapter) {
+  public MethodSpec createJsonAdapterMethod(TypeName autoValueClass,
+      TypeVariableName[] genericTypeNames, TypeSpec jsonAdapter) {
     ParameterSpec moshi = ParameterSpec.builder(Moshi.class, "moshi").build();
-    return MethodSpec.methodBuilder("jsonAdapter")
+    ParameterSpec type = ParameterSpec.builder(Type[].class, "types").build();
+    MethodSpec.Builder builder = MethodSpec.methodBuilder("jsonAdapter")
         .addModifiers(PUBLIC, STATIC)
         .addParameter(moshi)
-        .returns(ParameterizedTypeName.get(ClassName.get(JsonAdapter.class), autoValueClass))
-        .addStatement("return new $N($N)", jsonAdapter, moshi)
-        .build();
+        .returns(ParameterizedTypeName.get(ClassName.get(JsonAdapter.class), autoValueClass));
+
+    if(genericTypeNames != null) {
+      builder.addParameter(type)
+          .addTypeVariables(Arrays.asList(genericTypeNames))
+          .addStatement("return new $N($N, $N)", jsonAdapter, moshi, type);
+    }else
+      builder.addStatement("return new $N($N)", jsonAdapter, moshi);
+
+    return builder.build();
   }
 
-  public TypeSpec createTypeAdapter(ClassName className, ClassName autoValueClassName,
-      List<Property> properties) {
+  public TypeSpec createTypeAdapter(ClassName className, TypeName autoValueClassName,
+      TypeVariableName[] genericTypeNames, List<Property> properties) {
     ClassName jsonAdapter = ClassName.get(JsonAdapter.class);
     TypeName typeAdapterClass = ParameterizedTypeName.get(jsonAdapter, autoValueClassName);
 
     ImmutableMap<Property, FieldSpec> adapters = createFields(properties);
 
     ParameterSpec moshi = ParameterSpec.builder(Moshi.class, "moshi").build();
+    ParameterSpec type = ParameterSpec.builder(Type[].class, "types").build();
```

Comment:
`ParameterSpec type` is created to early... at this point we don't know if we even need it.